### ISSUE TITLE
Translating Tx Context to the Tx Context structure expected by a Plutus script generated in that era

### ISSUE THREAD

**Issue by polinavino** [IntersectMBO/cardano-ledger#1772]
We want to keep track of the slot numbers referred to in a script/datum/redeemer as slots in a given *ERA* (we need this for the complete information about an actual point in time that the contract is referring to) aligned with the era the transaction spending the given UTxO, so that the validity interval of a transaction refers to the same points in time that the contract thinks it is referring to. 

Since transactions will (likely) have an Era tag in the future, to be able to deserialize transactions from past eras, using this tag is a potential way to achieve this goal of making sure everything refers to the right eras.

**Comment by michaelpj** [comment]
I think this problem is a bit more general. Let me expand.

----

Recall that the type of a validator script is (morally) `Datum -> Redeemer -> Context -> Bool`.

The `Context` contains a representation of the transaction which is trying to spend the output. This can be difficult if the spending transaction is from a later era than when the output was created, for several reasons:
1. The representation of transactions may change in future eras (e.g. they may gain additional fields)
2. Some parts of transactions may actually mean something different in future eras (e.g. slot lengths could change). This might cause scripts to make incorrect decisions based on the old meanings.

However, we do need to be able to run old scripts in new eras, otherwise we will render many outputs unspendable.

Problem 1 already suggests that we need a way to "back-translate" a transaction into a transaction from the era that the script understands. Problem 2 extends the scope of this to say that we should *also* translate any parts of the transaction whose meaning has changed (e.g. translate slot numbers in validity intervals into the slot numbers that a ledger of the appropriate era would understand).

----

How do we know what era to back-translate to? This is the same as asking: what interface does the script expose? This second question is a bit more general, because the interface to the script is a bit more than just the representation of the transaction that gets passed in. For example, the `Context` for validator scripts also includes the index of the input currently being validated. So we could change this interface without *transactions* changing, and it would be equally problematic.

When we discussed this before, I think we concluded that the easiest way to handle this "interface version" was to just identify it with the language version. So `plc-v1` scripts will *always* expect to be given a representation of a transaction that looks like a Goguen transaction, and if we want to change that we will need a new version of the language. And we decided to do language versions by just adding new languages, so this amounts to adding a new `plc-v2` language.

This is likely to cause language-version (and hence new-language) churn, since even to *allow* people to write scripts that can see new fields of transactions we'll need a new language. So maybe we should reconsider this, and try and figure out some way of having an "interface version" that we can change independently.

----

Summing up:
- We are definitely going to need something like this back-translation functionality. I suspect this is going to be a problem, since AFAIK things are set up to do forwards-translation but not the reverse.
- We should think about language and interface versioning again.

**Comment by polinavino** [comment]
I think we are already making a function  (where the Tx is the Tx type of the current era)
`valContext ∈ UTxO → Tx → CurItem → Data`

It would make sense to pattern-match on the Era tag of the Tx (if that is indeed going to be part of Tx data in future eras), and also pass the language version of the script to this function, and use these pieces of info to generate the appropriate `Data` term. This would require some back-translating type operation, but I think since future eras will only add functionality, that should not be a problem (just drop all the irrelevant bits of data)

**Comment by michaelpj** [comment]
Yes, that's what this would amount to. I think it would be sensible to try and get such a back-translation function included in the "signature" for an era. Like you argued for having the forwards-translation one include. But this is even more important: if we can't forwards-translate, we won't be able to apply old transactions, but if we can't back-translate, we can't *spend* old outputs!

**Comment by polinavino** [comment]
Yeah - a back translation function would be specific to the era parameter of the ledger. Definitely needed.

**Comment by michaelpj** [comment]
I'm not sure I quite follow? Surely just `TxEra2 -> TxEra1`?

**Comment by polinavino** [comment]
I meant something like 
`valContext ∈ UTxO → Tx → CurItem → Era → Data` 

where `Tx inputs outputs ... originalSerializedTx originalTxEra` is a transaction already translated to the current format (ie Goguen Tx if it's currently the Goguen era, even if it was a Shelley Tx that came in)

and we have composable translation functions from each era to the next, so that whatever the current era is, any transaction from a past era (from Shelley onwards) can be translated to it

**Comment by michaelpj** [comment]
Right, so we're going to have to do something dumb like:
1. Get a transaction from EraN
2. Translate it to EraM (M>N) to pass to the ledger
3. Translate it back to EraN for a script that expects a EraN transaction

**Comment by polinavino** [comment]
Thaaaaat sounds like what I was saying, yes. We could try for something more elegant..?

**Comment by polinavino** [comment]
Done
